core/src/main/scala/kafka/server/instrumentation/ProduceRequestInstrumentation.scala

```diff
@@ -0,0 +1,193 @@
+package kafka.server.instrumentation
+
+import com.typesafe.scalalogging.Logger
+import kafka.cluster.PendingShrinkIsr
+import kafka.network.RequestChannel
+import kafka.server.instrumentation.ProduceRequestInstrumentation.Stage
+import kafka.server.instrumentation.ProduceRequestInstrumentationLogger.Config
+import kafka.server.{BrokerReconfigurable, KafkaConfig, ReplicaManager}
+import kafka.utils.CoreUtils.nanosToMs
+import kafka.utils.Logging
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.config.ConfigException
+import org.apache.kafka.common.utils.Time
+
+import scala.collection.mutable
+
+object ProduceRequestInstrumentation {
+  /**
+   * Symbols to be used for marking at each instrumentation stage
+   */
+  object Stage extends Enumeration {
```

Comment:
In Request.updateRequestMetrics, it updates the existing request latency metrics. I think it would be better to also cover those latencies in logs as well. Though we already have those in metrics, they are shown as P99/P99.9, not the exact number. Among those, apiRemoteTimeMs is one I think specifically important to have because for long latency request, this is the one is significant high. I do not think the current Stage cover this latency.

Reply:
Included those information.

@groelofs @hshi2022 also updated the PR description:

The updated log now looks like this
```
[2023-06-30 03:17:58,705] INFO acks=-1; request_metric={"responseQueueTimeMs":0,"responseBytes":56,"responseSendTimeMs":0,"requestQueueTimeMs":1,"sizeOfBodyInBytes":97,"apiRemoteTimeMs":17,"apiThrottleTimeMs":0,"totalTimeMs":38,"apiLocalTimeMs":20}; topic_partitions=(test1-0); stage_breakdown_ms={"Init: 1", "Authorization: 1", "BeginAppendRecords: 1", "AppendToLocalLog: 11", "ProcessAppendToLocalLogStatus: 2", "PrepareDelayedProduce: 0", "EnqueueDelayedProduce: 18", "BeginResponseCallback: 1", "ResponseThrottling: 2"}; PendingShrinkIsr_count=0 (produce.request.instrumentation.logger)
[2023-06-30 03:20:02,371] TRACE Funneled out by random coin 0.8470685658021838 > configured ratio 0.001 (produce.request.instrumentation.logger)
```